portfolio/src/main/java/com/google/sps/servlets/BusinessCardServlet.java

```diff
@@ -0,0 +1,186 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * This class will handle the storage/retrieval of information from the
+ * Business Card Drop on the portfolio homepage.
+ */
+@WebServlet("/biz-card")
+public class BusinessCardServlet extends HttpServletWithUtilities {
+
+  // Store the BlobstoreService instance for the application. Same instance as other files.
+  private final BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+
+  // Store the Datastore Service instance. Will hold all BizCard entities.
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  // Store the ImagesService instance which will be used to get download URLs for the images
+  // in blobstore.
+  private final ImagesService imagesService = ImagesServiceFactory.getImagesService();
+
+  /**
+   * Add a business card to the Datastore.
+   * DO NOT ACCESS DIRECTLY - should only be accessed by Blobstore.
+   * Point Client to the blobstore upload URL, which can be retrieved from the BlobstoreUrlServlet
+   *
+   * Image must have been uploaded inside a form element with the name "bizCard".
+   * If this changes, change the formElementName parameter in the getImageBlobKey function.
```

Comment:
Hmm...I'm actually not sure what the ideal implementation is. When the HTML form is sent, the name of any element becomes the Key and then the contents become the value. No matter what way I spin it, the Server has to query a key as an expected value. By changing the form element name, you would essentially be changing the key. 

I could create a post request on load that has the form element name as the value, but then that just transfers the problem to the key of that request. What if I change the name on the key of the new request? That problem would propagate forever. 

The `name` property of an HTML input element does not change how the form is accessed or used on the front end. CSS binds to element tags, class names, or ids. HTML selects elements based on those too. the `name` property is exclusively reserved as the "key" value in a request. There would be no reason for someone to change the name of the form element, and if they did I feel like failure in the server is an expected behaviour. 

I can't think of any changes that are worth the overhead/complexity they would add to the program. Please let me know if there's something you're thinking of - I'm curious if there's some cool feature/method I'm overlooking!

Reply:
I'm also changing these comments since this isn't necessarily true / possibly misleading. I could handle the upload in JavaScript and give the parameter any name I'd like and completely bypass the name of the form element.

Reply:
Good analysis! You are correct it probably doesn't matter in the context of this project. This is really an issue with any form that you post directly to the server, and someone changing the HTML code should know that field names affect the backend.

Larger projects tend to use a framework that dynamically generates HTML forms and also handles the results of submitting the form. These frameworks would generate the field name from either a constant or based on some other value (like a database column name).

You could potentially write a servelet that returns the desired field name, and use javascript to update the name field, but it would create more complexity than you need here.